jsonpath: add jsonpath conditional evaluation

normanchenn opened the pull request:
#### jsonpath: add jsonpath conditional evaluation

This commit adds support for evaluating jsonpath conditionals.
Conditionals have the structure of `$ ? (predicate)`, and will filter
the current json objects based on whether they satisfy the predicate
expression. Additionally, this commit introduces `@`, which represents
the current json object so filters can be used to reference the object
being evaluated.

Informs: #142610
Epic: None
Release note (sql change): Add jsonpath filters, which take on the form
`$ ? (predicate)`, allowing results to be filtered.

#### jsonpath/eval: add `unwrap` parameter to `jsonpath.eval`

Previously, jsonpath conditionals/filters would unwrap infinitely. For
example, `SELECT jsonb_path_query('{"a": [[[[{"b": 1}]]]]}', '$.a ? (@.b == 1)');`
would return an entry, when postgres would not. This commit adds an
unwrap boolean parameter to `eval`, which allows for control over when
to stop unwrapping json arrays. Additionally, this commit updates
`eval` to take in a `json.JSON` rather than a `[]json.JSON`, simplifying
the evaluation logic.

Epic: None
Release note: None

normanchenn committed:
jsonpath: add jsonpath conditional evaluation

This commit adds support for evaluating jsonpath conditionals.
Conditionals have the structure of `$ ? (predicate)`, and will filter
the current json objects based on whether they satisfy the predicate
expression. Additionally, this commit introduces `@`, which represents
the current json object so filters can be used to reference the object
being evaluated.

Epic: None
Release note (sql change): Add jsonpath filters, which take on the form
`$ ? (predicate)`, allowing results to be filtered.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143097)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
Very nice! I only have a handful of nits.

> Reviewed 9 of 9 files at r1, 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143097#-:-OLjrM9V9AXbFsj-nNar:bmw6hso)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @normanchenn)

___
*[`pkg/util/jsonpath/eval/array.go` line 19 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143097#-OLjqL2fE80LOa0ZdAoo:-OLjqL2fE80LOa0ZdAop:b-kzolwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2af3fdc4c5fa1622708448e1e204e0ccd6eb6cc/pkg/util/jsonpath/eval/array.go#L19)):*
> ```Go
> 	if jsonValue.Type() == json.ArrayJSONType {
> 		// Do not evaluate any paths, just unwrap the current target.
> 		return ctx.unwrapCurrentTargetAndEval(nil, jsonValue, !ctx.strict)
> ```

nit: add `/* jsonPath */` after `nil`.
___
*[`pkg/util/jsonpath/eval/array.go` line 72 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143097#-OLjoZoO9lCfeQhZeHum:-OLjoZoO9lCfeQhZeHun:bzeu5lk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2af3fdc4c5fa1622708448e1e204e0ccd6eb6cc/pkg/util/jsonpath/eval/array.go#L72)):*
> ```Go
> 		}
> 		return agg, nil
> 	} else {
> ```

nit: with error cases like this it's a bit cleaner to do the negative check first, return an error, and then have the happy logic outside of  the `if` block, i.e. something like
```
if jsonValue.Type() != json.ArrayJSONType && ctx.strict {
  < return error >
}
< do the happy path logic >
```
___
*[`pkg/util/jsonpath/parser/testdata/jsonpath` line 359 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143097#-OLjnq784W7G77KsKOrj:-OLjnq784W7G77KsKOrk:b-vm4c17) ([raw file](https://github.com/cockroachdb/cockroach/blob/1157d89db49d780e1e600ac7be822f49c3bd45c9/pkg/util/jsonpath/parser/testdata/jsonpath#L359)):*
> ```
> $.a[*] ? (@.b > 100)
> ----
> $."a"[*]?((@."b" > 100)) -- normalized!
> ```

nit: the postgres always uses single parenthesis (even if the original had multiple), should we leave a TODO for this here?
___
*[`pkg/util/jsonpath/eval/operation.go` line 141 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143097#-OLjq5pH4vps9nAB9BZC:-OLjq5pH4vps9nAB9BZD:b40x8xd) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2af3fdc4c5fa1622708448e1e204e0ccd6eb6cc/pkg/util/jsonpath/eval/operation.go#L141)):*
> ```Go
> 	op jsonpath.Operation, jsonValue json.JSON, unwrapRight bool,
> ) (jsonpathBool, error) {
> 	left, err := ctx.evalAndUnwrapResult(op.Left, jsonValue, true)
> ```

nit: a quick comment for why we're always unwrapping on the left would be helpful.
___
*[`pkg/util/jsonpath/eval/filter.go` line 17 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143097#-OLjpGeh1SgRpmK7pEyQ:-OLjpGeh1SgRpmK7pEyR:b970wa0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2af3fdc4c5fa1622708448e1e204e0ccd6eb6cc/pkg/util/jsonpath/eval/filter.go#L17)):*
> ```Go
> ) ([]json.JSON, error) {
> 	if unwrap && current.Type() == json.ArrayJSONType {
> 		return ctx.unwrapCurrentTargetAndEval(p, current, false)
> ```

nit: add `/* unwrap */` after `false`.

It is a common convention to specify the name of parameters when unnamed constants (like `false` or `nil`) are passed in the production code. Additionally, it might also be helpful to do the same after `!ctx.strict` in all callsites, to make it more clear too.
___
*[`pkg/util/jsonpath/eval/eval.go` line 122 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143097#-OLjqoNBAjy-uFR3XIYb:-OLjqoNBAjy-uFR3XIYc:b2i5r2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2af3fdc4c5fa1622708448e1e204e0ccd6eb6cc/pkg/util/jsonpath/eval/eval.go#L122)):*
> ```Go
> ) ([]json.JSON, error) {
> 	if jsonValue.Type() != json.ArrayJSONType {
> 		return nil, errors.Newf("unwrapCurrentTargetAndEval can only be applied to an array")
> ```

nit: should `errors.Newf` be assertion failures here and in `executeAnyItem`?


<!-- Sent from Reviewable.io -->

normanchenn committed:
jsonpath/eval: add unwrap parameter to jsonpath.eval

Previously, jsonpath conditionals/filters would unwrap infinitely. For
example, `SELECT jsonb_path_query('{"a": [[[[{"b": 1}]]]]}', '$.a ? (@.b == 1)');`
would return an entry, when postgres would not. This commit adds an
unwrap boolean parameter to `eval`, which allows for control over when
to stop unwrapping json arrays. Additionally, this commit updates
`eval` to take in a `json.JSON` rather than a `[]json.JSON`, simplifying
the evaluation logic.

Epic: None
Release note: None

normanchenn reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143097)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/util/jsonpath/parser/testdata/jsonpath` line 359 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143097#-OLjnq784W7G77KsKOrj:-OLkEQQd9QraMcCBHc0_:bvketdb) ([raw file](https://github.com/cockroachdb/cockroach/blob/1157d89db49d780e1e600ac7be822f49c3bd45c9/pkg/util/jsonpath/parser/testdata/jsonpath#L359)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: the postgres always uses single parenthesis (even if the original had multiple), should we leave a TODO for this here?
</blockquote></details>

Yep, I added a TODO in `pkg/util/jsonpath/operation.go:Operation.String()` about this in a previous commit.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
Thanks! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 7 of 7 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143097#-:-OLkPc874el-uqdyXQZU:b-ftzx47)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @normanchenn)



<!-- Sent from Reviewable.io -->

normanchenn commented:
TFTR!

bors r+
